components/raftstore/src/store/fsm/peer.rs

```diff
@@ -753,157 +743,159 @@ where
-            .collect();
-        if to_be_removed.is_empty()
-            && self.region().get_start_key() == region.get_start_key()
-            && self.region().get_end_key() == region.get_end_key()
-        {
-            // Nothing to be updated, return directly.
+    fn on_unsafe_recovery_pre_demote_failed_voters(
+        &mut self,
+        shared_state: UnsafeRecoveryExecutePlanSyncer,
+        failed_voters: Vec<metapb::Peer>,
+    ) {
+        if !self.fsm.peer.is_force_leader() {
+            error!(
+                "Unsafe recovery, demoting failed voters failed, since this peer is not forced leader";
+                "region" => ?self.region(),
+            );
             return;
         }
-        info!(
-            "updating the reigon for unsafe recover, original: {:?}, target: {:?}",
-            self.region(),
-            region
-        );
-        if self.fsm.peer.has_valid_leader() {
-            panic!("region update for unsafe recover should only occur in leaderless reigons");
-        }
-        if self.fsm.peer.raft_group.store().applied_index()
-            != self.fsm.peer.raft_group.store().commit_index()
-        {
+
+        if self.fsm.peer.unsafe_recovery_state.is_some() {
             warn!(
-                "cannot proceed region update for unsafe recover, applied index is not equal to commit index"
+                "Unsafe recovery, demote failed voters has already been initiated";
+                "region" => ?self.region(),
             );
             return;
         }
 
-        let region_state_key = keys::region_state_key(region.get_id());
-        let original_region_state = match self
-            .ctx
-            .engines
-            .kv
-            .get_msg_cf::<RegionLocalState>(CF_RAFT, &region_state_key)
-        {
-            Ok(Some(region_state)) => region_state,
-            Ok(None) => {
-                panic!("Can't find RegionLocalState while updating {:?}", region);
-            }
-            Err(e) => {
-                panic!(
-                    "Fail to look up RegionLocalState while updating {:?} err {:?}",
-                    region, e
-                );
+        if self.fsm.peer.in_joint_state() {
+            info!(
+                "Unsafe recovery, already in joint state, exit first";
+                "region" => ?self.region());
+            let proposed = Arc::new(Mutex::new(false));
+            let proposed_clone = proposed.clone();
+            let callback = Callback::<EK::Snapshot>::write_ext(
+                Box::new(|resp| {
+                    if resp.response.get_header().has_error() {
+                        error!(
+                            "Unsafe recovery, fail to exit residual joint state";
+                            "err" => ?resp.response.get_header().get_error(),
+                        );
+                    }
+                }),
+                Some(Box::new(move || {
+                    *proposed_clone.lock().unwrap() = true;
```

Comment:
seems it isn't called everytime propose succeed

Reply:
Refactored.

Now we share a `failed` bool between the caller and the write callback, if the `failed` is not set to true when the caller examines it, proposal is at least accepted by Raft, thus we can set `unsafe recovery state` accordingly.